### DISCUSSION TITLE
Fabric and Visual Studio Code Windows

### DISCUSSION THREAD

**Discussion by dsaldivar** [danielmiessler/Fabric#124] [Show and tell]
Just an update of the saga of my trying to do the installation of fabric using Visual Studio Code in Powershell terminal. To make life easier for yourself and be able to follow the Readme.md instructions verbatim you need to install the extension Windows Subsystem for Linux. To do this in Visual Studio Code go to File \ Preferences \ Extensions and look for it there. Now I'm not sure if you first need to turn this on as an App Feature within Windows first before seeing it on the list of Extensions in Visual Studio Code but if you don't see WSL as an optional Extension then proceed to follow the instructions to do that in Windows. At google just type in turning on Windows Subsystem For Linux App Feature in Windows and follow the instructions. Pretty straight forward. Again I'm not sure if you need to do this outside of Visual Studio Code or not before it becoming an available extension. Just a precaution if necessary. Anyway after you have done that you can go back into Visual Studio Code and open a new terminal. In the window of the terminal screen there is a small down arrow that when clicked on will show a menu of other terminal environments you can use. One should be there called Ubuntu (WSL) use this terminal environment and then you can begin following the instructions as outlined in Readme.md. Just so we are clear, I'm not sure if these instructions will actually work for you on you system because of the myriad of setup nuances that may exist on everyone's Windows environment. But the beauty is you have an excellent friend in Chatgpt to help guide you. Hope this helps some of you.

**Comment by rupjae** [comment]
The above set me on the right path and, I didn't even know about WSL which is even better because it is quite awesome.
In case others are going down this path and noticed slow network connectivity when downloading ubuntu or updating the os repo have a look at this article on how to change the virtual interface. 
https://townsyio.medium.com/wsl2-how-to-fix-download-speed-3edb0c348e29

**Comment by danielmiessler** [comment]
Very cool!

**Comment by agathaaaaa** [comment]
Any other way to install it other than using WSL?

**Reply by ksylvan** [reply]
Yes. You can install Fabric in native windows mode.

**Comment by Gerkinfeltser** [comment]
I was able to get Fabric up and running via WSL. I'm curious what other folks use in place of `pbpaste`. So far I'm just making files can doing `cat thisfile.md | fabric --pattern summarize`. I'd love to hear alternatives though.

**Reply by wally-kroeker** [reply]
This worked for me.  Simple alias 
https://dev.to/emmiep/better-copy-and-paste-in-wsl-with-two-one-liner-aliases-ldm

**Reply by wally-kroeker** [reply]
alias pbcopy='powershell.exe -Command "Set-Clipboard -Value \$input"'
alias pbpaste='powershell.exe -Command "Get-Clipboard"'

**Comment by Gerkinfeltser** [comment]
Another follow up from me. I was able to get fabric working in a vs code terminal without WSL. I haven't done much testing on it but so far it seems to be working comparably well. I haven't been thru the pull process yet though to update. Hopefully that won't cause issues.

I posted what I did to get it working here: #183 (comment)

**Reply by Gerkinfeltser** [reply]
Ah, one issue so far is that `~` doesn't mean the same thing in Windows as it does in Linux/Mac. What's interesting is that a .config/fabric  dir and .env got created (even so far as getting the correct openai key) but doing things like `fabric --changeDefaultModel doesn't work.

**Comment by ksylvan** [comment]
I have OLLAMA working on the Windows host and a perfectly running fabric on the WSL Linux Ubuntu host, now how do I get the two of them talking?

```
kayvan@SHAKTI:/mnt/c/Users/kayvan/src/fabric$ ollama.exe list
NAME                    ID              SIZE    MODIFIED    
llama2:latest           78e26419b446    3.8 GB  2 weeks ago
mistral:latest          61e88e884507    4.1 GB  2 hours ago
mixtral:instruct        7708c059a8bb    26 GB   2 hours ago
mixtral:latest          7708c059a8bb    26 GB   2 weeks ago

kayvan@SHAKTI:/mnt/c/Users/kayvan/src/fabric$ fabric --listmodels
GPT Models:
gpt-3.5-turbo
gpt-3.5-turbo-0125
gpt-3.5-turbo-0301
gpt-3.5-turbo-0613
gpt-3.5-turbo-1106
gpt-3.5-turbo-16k
gpt-3.5-turbo-16k-0613
gpt-3.5-turbo-instruct
gpt-3.5-turbo-instruct-0914
gpt-4
gpt-4-0125-preview
gpt-4-0613
gpt-4-1106-preview
gpt-4-turbo-preview
gpt-4-vision-preview

Local Models:

Claude Models:
claude-3-opus-20240229
claude-3-sonnet-20240229
claude-3-haiku-20240307
claude-2.1
```

**Comment by ksylvan** [comment]
Following up to my comment above, I read the Ollama issue here: https://github.com/ollama/ollama/issues/703

In PowerShell, I do this:

```
PS C:\Users\kayvan> $env:OLLAMA_HOST="0.0.0.0"
PS C:\Users\kayvan> ollama serve
time=2024-03-18T06:57:00.504-07:00 level=INFO source=images.go:806 msg="total blobs: 14"
time=2024-03-18T06:57:00.504-07:00 level=INFO source=images.go:813 msg="total unused blobs removed: 0"
time=2024-03-18T06:57:00.505-07:00 level=INFO source=routes.go:1110 msg="Listening on [::]:11434 (version 0.1.29)"
time=2024-03-18T06:57:00.505-07:00 level=INFO source=payload_common.go:112 msg="Extracting dynamic libraries to C:\\Users\\kayvan\\AppData\\Local\\Temp\\ollama3941994260\\runners ..."
time=2024-03-18T06:57:00.561-07:00 level=INFO source=payload_common.go:139 msg="Dynamic LLM libraries [rocm_v5.7 cpu_avx2 cuda_v11.3 cpu_avx cpu]"
```

And now in the WSL Ubuntu environment, I can see this:

```
kayvan@SHAKTI:/mnt/c/Users/kayvan/src/fabric$ ollama.exe list
NAME                    ID              SIZE    MODIFIED     
llama2:latest           78e26419b446    3.8 GB  2 weeks ago 
mistral:latest          61e88e884507    4.1 GB  13 hours ago
mixtral:instruct        7708c059a8bb    26 GB   13 hours ago
mixtral:latest          7708c059a8bb    26 GB   2 weeks ago
```

And I can now see this, using the default gateway:

```
kayvan@SHAKTI:/mnt/c/Users/kayvan/src/fabric$ cat /etc/resolv.conf 
# This file was automatically generated by WSL. To stop automatic generation of this file, add the following entry to /etc/wsl.conf:
# [network]
# generateResolvConf = false
nameserver 172.23.112.1

kayvan@SHAKTI:/mnt/c/Users/kayvan/src/fabric$ curl http://172.23.112.1:11434; echo ""
Ollama is running
```

But I'm still not seeing the local models in fabric:

```
$ fabric --remoteOllamaServer http://172.23.112.1:11434 --listmodels
GPT Models:
gpt-3.5-turbo
gpt-3.5-turbo-0125
gpt-3.5-turbo-0301
gpt-3.5-turbo-0613
gpt-3.5-turbo-1106
gpt-3.5-turbo-16k
gpt-3.5-turbo-16k-0613
gpt-3.5-turbo-instruct
gpt-3.5-turbo-instruct-0914
gpt-4
gpt-4-0125-preview
gpt-4-0613
gpt-4-1106-preview
gpt-4-turbo-preview
gpt-4-vision-preview

Local Models:

Claude Models:
claude-3-opus-20240229
claude-3-sonnet-20240229
claude-3-haiku-20240307
claude-2.1
```

Any ideas for how to get local models working in Windows/WSL?
